builtin/logical/database/backend.go

```diff
@@ -166,46 +140,138 @@ func (b *databaseBackend) Role(ctx context.Context, s logical.Storage, roleName
 		return nil, err
 	}
 
-	empty := upgradeCheck{}
-	if upgradeCh != empty {
-		result.Statements.CreationStatements = upgradeCh.Statements.CreationStatements
-		result.Statements.RevocationStatements = upgradeCh.Statements.RevocationStatements
-		result.Statements.RollbackStatements = upgradeCh.Statements.RollbackStatements
-		result.Statements.RenewStatements = upgradeCh.Statements.RenewStatements
+	switch {
+	case upgradeCh.Statements != nil:
+		var stmts dbplugin.Statements
+		if upgradeCh.Statements.CreationStatements != "" {
+			stmts.Creation = []string{upgradeCh.Statements.CreationStatements}
+		}
+		if upgradeCh.Statements.RevocationStatements != "" {
+			stmts.Revocation = []string{upgradeCh.Statements.RevocationStatements}
+		}
+		if upgradeCh.Statements.RollbackStatements != "" {
+			stmts.Rollback = []string{upgradeCh.Statements.RollbackStatements}
+		}
+		if upgradeCh.Statements.RenewStatements != "" {
+			stmts.Renewal = []string{upgradeCh.Statements.RenewStatements}
+		}
+		result.Statements = stmts
 	}
 
+	// For backwards compatibility, copy the values back into the string form
+	// of the fields
+	result.Statements = dbutil.StatementCompatibilityHelper(result.Statements)
+
 	return &result, nil
 }
 
 func (b *databaseBackend) invalidate(ctx context.Context, key string) {
-	b.Lock()
-	defer b.Unlock()
-
 	switch {
 	case strings.HasPrefix(key, databaseConfigPath):
 		name := strings.TrimPrefix(key, databaseConfigPath)
-		b.clearConnection(name)
+		b.ClearConnection(name)
 	}
 }
 
-// clearConnection closes the database connection and
+func (b *databaseBackend) GetConnection(ctx context.Context, s logical.Storage, name string) (*dbPluginInstance, error) {
+	b.RLock()
+	unlockFunc := b.RUnlock
+	defer func() { unlockFunc() }()
+
+	db, ok := b.connections[name]
+	if ok {
+		return db, nil
+	}
+
+	// Upgrade lock
+	b.RUnlock()
+	b.Lock()
+	unlockFunc = b.Unlock
+
+	db, ok = b.connections[name]
+	if ok {
+		return db, nil
+	}
+
+	config, err := b.DatabaseConfig(ctx, s, name)
+	if err != nil {
+		return nil, err
+	}
+
+	dbp, err := dbplugin.PluginFactory(ctx, config.PluginName, b.System(), b.logger)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = dbp.Init(ctx, config.ConnectionDetails, true)
+	if err != nil {
+		dbp.Close()
+		return nil, err
+	}
+
+	id, err := uuid.GenerateUUID()
+	if err != nil {
+		return nil, err
+	}
+
+	db = &dbPluginInstance{
+		Database: dbp,
+		name:     name,
+		id:       id,
+	}
+
+	b.connections[name] = db
+	return db, nil
+}
+
+// ClearConnection closes the database connection and
 // removes it from the b.connections map.
-func (b *databaseBackend) clearConnection(name string) {
+func (b *databaseBackend) ClearConnection(name string) error {
+	b.Lock()
+	defer b.Unlock()
+	return b.clearConnection(name)
+}
+
+func (b *databaseBackend) clearConnection(name string) error {
 	db, ok := b.connections[name]
 	if ok {
-		db.Close()
+		if err := db.Close(); err != nil {
```

Comment:
I think we will want to clear the connection here even if there is an error. The error comes from the transport call to the plugin, but the plugin is always killed when Close is called: https://github.com/hashicorp/vault/blob/master/builtin/logical/database/dbplugin/client.go#L26